src/compile/data/optimizers.ts

```diff
@@ -15,10 +15,10 @@ import {CalculateNode, FilterNode} from './transforms';
  */
 export function optimizeFromLeaves(f: (node: DataFlowNode) => void) {
   function optimizeNextFromLeaves(node: DataFlowNode) {
-    if (node.parent instanceof SourceNode || node instanceof SourceNode) {
+    if (node.parent instanceof SourceNode) {
       return;
     } else if (!node || !node.parent) {
-      throw new Error('A source node cannot have parents and roots haev to be source nodes.');
+      throw new Error('Bad state: a source node cannot have parents and roots have to be source nodes.');
```

Comment:
I still think that this error is very confusing. 

In this branch you can only conclude that
1. Node.parent is not a SourceNode.
2. Either `!node` or `!node.parent`.
For the former one (!node), it's very weird why would that case happen at all.
For the latter (`!node.parent`), yes this node is a root.  While you can infer that root should be a source node (comment missing), you cannot say that this source node has parents.